[scroll-animations] address WPT failures related to `timeline-scope` and `animation-timeline`

graouts opened the pull request:
#### dd9fc65cb4bc5cd6455330e6ae0cca0d9cfe328d
<pre>
[scroll-animations] address WPT failures related to `timeline-scope` and `animation-timeline`
<a href="https://bugs.webkit.org/show_bug.cgi?id=287337">https://bugs.webkit.org/show_bug.cgi?id=287337</a>

Reviewed by Anne van Kesteren.

In 290046@main we put together a sounder approach to attaching timelines to style-originated
animations by waiting until the style resolution process completed. But the attachment logic
as well as when to attempt attachment immediately or defer it was lacking and resulted in many
WPT failures related to the `timeline-scope` and `animation-timeline` properties.

We now add a parameter to `AnimationTimelinesController::setTimelineForName()` that determines
whether attachment deferral until style resolution has completed is allowed (ie. when we&apos;re
processing the `animation-timeline` property during style resolution) or to attempt attachment
immediately (ie. when style resolution has completed).

Then when we enter `setTimelineForName()`, we have three options:

1. If deferral is allowed and we don&apos;t have an active named timeline, wait until style resolution
has completed to see if a timeline exists then.

2. If deferral is not allowed and we still don&apos;t have an active named timeline, assign either a
placeholder inactive scroll timeline if the name is in scope, or a null timeline otherwise.

3. If we have an active named timeline, determine which one to use per `determineTimelineForElement()`.

Finally, when the `animation-timeline` property changes to a value that is not a timeline name – such
as `auto`, `none`, `scroll()` or `view()` – we remove all pending attachment operations for this CSS
Animation.

There remain a few WPT failures but these will be handled by future individual patches.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-deferred-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-named-scroll-progress-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-dynamic.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::removePendingOperationsForCSSAnimation):
(WebCore::AnimationTimelinesController::documentDidResolveStyle):
(WebCore::AnimationTimelinesController::unregisterNamedTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):

Canonical link: <a href="https://commits.webkit.org/290145@main">https://commits.webkit.org/290145@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39e3cb5f25bbbc36d967ed9a91a205ab8d05db09

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88975 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8499 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43930 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93947 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39735 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91026 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8886 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16683 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68549 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26224 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6779 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/80870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48913 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6528 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38843 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76890 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36233 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95786 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16155 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11798 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77427 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16411 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76715 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21102 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9238 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13964 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16169 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21480 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15910 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19361 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17691 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -589,20 +603,50 @@ void AnimationTimelinesController::setTimelineForName(const AtomString& name, co
+    } else {
+        auto& timelines = it->value;
+        if (RefPtr timeline = determineTimelineForElement(timelines, styleable, timelineScopeElements)) {
+            LOG_WITH_STREAM(Animations, stream << "AnimationTimelinesController::setTimelineForName: " << name << " styleable: " << styleable << " attaching to timeline of element: " << originatingElement(*timeline));
```
Is this meant to be included?

graouts replied:
It was existing code that was moved here (see above in the red block).

annevk commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -566,21 +569,32 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
         return;
 
-    auto& timeline = timelines.at(i);
+    auto timeline = timelines.at(i);
```
Note: in this PR a variable named `timeline` uses `auto`, `auto&`, and `RefPtr`. Perhaps that's reasonable, but it did strike me as a lil weird.

graouts replied:
Let's look at them one by one:

- `auto timeline` in `AnimationTimelinesController::unregisterNamedTimeline()`: this one creates an additional `Ref<>` for a timeline that will, in the next line, be removed from the vector and thus have its ref count decreased.
- `RefPtr timeline` to hold the return value for `determineTimelineForElement()` in `AnimationTimelinesController::setTimelineForName`: the method `determineTimelineForElement()` returns a raw pointer, so per our programming guidelines we create a `RefPtr<>` for this pointer.
- `auto& timeline` in `CSSAnimation::syncStyleOriginatedTimeline()`: this one holds return value for `Animation::timeline()`, which is just a non-ref-counted `Animation::Timeline` struct. That return value is reference.

So they all are as intended, the use of `auto` does make it a bit more difficult to understand what those values hold.

graouts committed:
[scroll-animations] address WPT failures related to `timeline-scope` and `animation-timeline`
https://bugs.webkit.org/show_bug.cgi?id=287337

Reviewed by Anne van Kesteren.

In 290046@main we put together a sounder approach to attaching timelines to style-originated
animations by waiting until the style resolution process completed. But the attachment logic
as well as when to attempt attachment immediately or defer it was lacking and resulted in many
WPT failures related to the `timeline-scope` and `animation-timeline` properties.

We now add a parameter to `AnimationTimelinesController::setTimelineForName()` that determines
whether attachment deferral until style resolution has completed is allowed (ie. when we're
processing the `animation-timeline` property during style resolution) or to attempt attachment
immediately (ie. when style resolution has completed).

Then when we enter `setTimelineForName()`, we have three options:

1. If deferral is allowed and we don't have an active named timeline, wait until style resolution
has completed to see if a timeline exists then.

2. If deferral is not allowed and we still don't have an active named timeline, assign either a
placeholder inactive scroll timeline if the name is in scope, or a null timeline otherwise.

3. If we have an active named timeline, determine which one to use per `determineTimelineForElement()`.

Finally, when the `animation-timeline` property changes to a value that is not a timeline name – such
as `auto`, `none`, `scroll()` or `view()` – we remove all pending attachment operations for this CSS
Animation.

There remain a few WPT failures but these will be handled by future individual patches.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-deferred-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-named-scroll-progress-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-dynamic.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::removePendingOperationsForCSSAnimation):
(WebCore::AnimationTimelinesController::documentDidResolveStyle):
(WebCore::AnimationTimelinesController::unregisterNamedTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):

Canonical link: https://commits.webkit.org/290145@main
